### ISSUE TITLE
auto-add import statements?

### ISSUE THREAD

**Issue by trusktr** [TypeStrong/atom-typescript#1455]
On `ide-typescript`, I was able to add import statements for things I'd forgotten to import. There doesn't seem to be this feature in atom-typescript (v12). Any chance of adding it?

**Comment by lierdakil** [comment]
We support this via code actions. Bascially hover over an unimported name and you should be presented with some options. This is a tsserver feature (meaning the core TypeScript team is working on it, we're just providing some GUI bindings)
![image](https://user-images.githubusercontent.com/7275622/46736966-00cd7300-cca3-11e8-8eb4-24ec8a5daa36.png)

(this can also be done using intentions Atom package insead of atom-ide-ui, if you want to; I personally prefer atom-ide-ui though)

**Comment by trusktr** [comment]
Oh, interesting! Okay I'll lookout for that box with that option.

In the project I'm working on, I don't ever seem to see that box. I only see the helpful blue boxes:

<img width="918" alt="screen shot 2018-10-29 at 2 03 58 pm" src="https://user-images.githubusercontent.com/297678/47680248-85fdc680-db83-11e8-8dea-a91dd8dd7f2c.png">

It seems I don't see type errors, therefore I never see that option.

Is there any reason you can think of why I may not ever see type errors?

**Comment by lierdakil** [comment]
That "box" is a feature of atom-ide-ui specifically. Kinda. Anyway. I personally recommend using atom-ide-ui as an all-around more "batteries included" option, but you can replicate the functionality with a few other packages, as outlined [in the readme](https://github.com/TypeStrong/atom-typescript#installation) ("installation", "dependencies" subsection)

If you don't want to use atom-ide-ui for whatever reason, you can use `intentions` package for a similar, although not entirely equivalent, effect: you can bring up intentions at text cursor via `intentions:show` command (bound to alt+enter by default I think)
![intentions menu for importing unresolved readFileSync import is shown](https://user-images.githubusercontent.com/7275622/47757722-c317a500-dcb8-11e8-840d-3a272870c79a.png)

You will also likely need linter (with default-ui) to show error positions etc.

**Comment by trusktr** [comment]
How do I "choose" atom-ide-ui? Do I have to explicitly disable the other packages?

**Comment by trusktr** [comment]
Is there something built into Atom that specifies that various packages can fulfill the role of the ui provider, based on which is installed? Or is it only convention among package authors cooperating with the provider packages?

**Comment by trusktr** [comment]
I see this every time I start up, but it never goes away (I've tried clicking on both buttons).

<img width="475" alt="screen shot 2018-11-04 at 11 03 16 am" src="https://user-images.githubusercontent.com/297678/47968603-68b47680-e021-11e8-8eb1-2abb754cb1e4.png">

Not sure if I should open a new issue, but I don't think I know what's going on in general to know if there's even a problem.

Once I get this all working well it may go away, and if not I can be more confident about reporting a problem if there is one.

**Comment by trusktr** [comment]
I tried disabling `intentions` and `linter-ui-default`, and left `atom-ide-ui` enabled hoping I'd get errors, etc, from that, but I get nothing. No errors anywhere (even though I purposefully make type errors).

The only thing I do see is the blue type tooltip when hovering:

<img width="327" alt="screen shot 2018-11-04 at 11 08 52 am" src="https://user-images.githubusercontent.com/297678/47968669-08720480-e022-11e8-8694-7d2b50ec8d65.png">

**Comment by trusktr** [comment]
Eh, Well, I can't seem to get `atom-ide-ui` to show anything at all for `atom-typescript`. `¯\_(ツ)_/¯`

And after experimenting with `linter-ui-default` and `linter-ui-plus`, looks like what I need to do is make my own taking the best features from both (underline files/folders and tooltips from `linter-ui-default` and super nice working panel from `linter-ui-plus`), and add gutter icon tooltips.

**Comment by lierdakil** [comment]
> Do the intentions work the same as the ones that I should see with atom-ide-ui?

Both share a lot of backend code, so yes, more or less, the functionality should be the same.

> I can't seem to get atom-ide-ui to show anything at all

Probably because you need to disable `linter` package, not just `linter-ui-*`. Also, make sure atom-ide-ui has diagnostics component enabled (i.e. check your config)

> Is there something built into Atom that specifies that various packages can fulfill the role of the ui provider, based on which is installed? Or is it only convention among package authors cooperating with the provider packages?

It's a convention. Atom doesn't bother with service arbitration, so package authors have to bother with that instead. Few ever actually do. So yeah, most linter-dependent packages will either only work with the first frontend they find, or will try to work with each they find (often with disastrous results). Hence, you *do not* want to have both `linter` (verbatim) and `atom-ide-ui` diagnostics component active at the same time.

The pop-up that prompts you to choose linter UI is a feature of atom-ide-ui. Not sure why the buttons don't work though. Might be a bug.

**Comment by trusktr** [comment]
Gotcha! These are little things that can only be discovered by tinkering for a while. Hopefully this helps someone who might stumble here.

I actually ended up with a custom configuration that is working quite well:

- I installed both `linter-ui-default` and `linter-ui-plus`
- I disabled the panel button for `linter-ui-default` in the settings, but kept the nice tooltips.
- `linter-ui-plus` provides nice gutter icons and panel. So to avoid having both gutter icons I set `display: none` to the gutter from `linter-ui-default` (see below)
- `linter-ui-default` provides the nice squiggly underlines on files/folders in the file explorer.
- Both libs draw squiggly underlines under code where errors/warnings/info happen, but the ones from `linter-ui-default` look better, so I set `display: none` to the ones from `linter-ui-plus` (see below)
- Finally minimap-linter for the minimap

Here's what I put in my Atom stylesheet:

```less
// hide the gutter from linter-ui-default:
[gutter-name='linter-ui-default'] {
	display: none;
}

// disable the squiggly underlines from linter-ui-plus:
.highlights {
	.highlight[class*='linter-highlight-'] {
		display: none;
	}
}
```

And it looks like follows with all features visible (except the blue type tips):

<img width="1792" alt="screen shot 2018-11-04 at 12 49 16 pm" src="https://user-images.githubusercontent.com/297678/48016692-19357f80-e0e1-11e8-8521-36d728118902.png">

And the blue type tooltips separately because they can overlap on top of the error tips:

<img width="1792" alt="screen shot 2018-11-04 at 12 51 50 pm" src="https://user-images.githubusercontent.com/297678/48016718-2b172280-e0e1-11e8-97fc-9a7dd5b3153a.png">

I wish the intentions were visible in the error/info tooltips, but that's for another day.
